File: scripts/download_gifs.py
#!/usr/bin/env python3
"""
Exercise GIF Downloader Script

Download exercise GIFs from URLs and save them with proper naming

Usage:
    python scripts/download_gifs.py
    
Or edit the EXERCISE_URLS list below and run
"""

import os
import sys
import urllib.request
import urllib.error
from pathlib import Path
import time

# ===== CONFIGURATION =====
# Add your GIF URLs here with exercise names
EXERCISE_URLS = [
    # Example format:
    # ("Bench Press", "https://example.com/bench-press.gif"),
    # ("Squat", "https://example.com/squat.gif"),
    # ("Deadlift", "https://example.com/deadlift.gif"),
    
    # Add your URLs below:
    
]

# Output directory (relative to project root)
SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
OUTPUT_DIR = PROJECT_ROOT / "public" / "exercise-gifs"

# ===== HELPER FUNCTIONS =====

def to_kebab_case(text):
    """Convert exercise name to kebab-case filename"""
    return text.lower().replace(' ', '-').replace('_', '-')

def download_file(url, filepath):
    """Download a file from URL with progress"""
    try:
        # Add headers to mimic a browser
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
        req = urllib.request.Request(url, headers=headers)
        
        with urllib.request.urlopen(req, timeout=30) as response:
            if response.status != 200:
                raise Exception(f"HTTP {response.status}")
            
            # Get file size if available
            total_size = response.headers.get('Content-Length')
            
            with open(filepath, 'wb') as f:
                f.write(response.read())
            
        return True
    except Exception as e:
        if filepath.exists():
            filepath.unlink()
        raise e

def format_size(size_bytes):
    """Format bytes to MB"""
    return f"{size_bytes / (1024 * 1024):.2f} MB"

def main():
    print("🎬 Exercise GIF Downloader\n")
    
    # Ensure output directory exists
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    print(f"✅ Output directory: {OUTPUT_DIR}\n")
    
    if not EXERCISE_URLS:
        print("⚠️  No URLs configured!\n")
        print("Please edit this file and add URLs to the EXERCISE_URLS list.\n")
        print("Example:")
        print('  ("Bench Press", "https://example.com/bench-press.gif"),')
        print('  ("Squat", "https://example.com/squat.gif"),\n')
        return
    
    print(f"📥 Downloading {len(EXERCISE_URLS)} exercise GIFs...\n")
    
    success_count = 0
    fail_count = 0
    skip_count = 0
    
    for name, url in EXERCISE_URLS:
        filename = f"{to_kebab_case(name)}.gif"
        filepath = OUTPUT_DIR / filename
        
        # Skip if file already exists
        if filepath.exists():
            size = filepath.stat().st_size
            print(f"  ⏭️  \"{name}\" → {filename}... Already exists ({format_size(size)})")
            skip_count += 1
            continue
        
        try:
            print(f"  Downloading \"{name}\"... ", end='', flush=True)
            download_file(url, filepath)
            
            size = filepath.stat().st_size
            print(f"✅ ({format_size(size)})")
            success_count += 1
            
            # Small delay to be nice to servers
            time.sleep(0.5)
            
        except Exception as e:
            print(f"❌ Failed: {e}")
            fail_count += 1
    
    print("\n" + "="*60)
    print(f"✅ Success: {success_count}")
    if skip_count > 0:
        print(f"⏭️  Skipped (already exist): {skip_count}")
    if fail_count > 0:
        print(f"❌ Failed: {fail_count}")
    print("="*60 + "\n")
    
    if success_count > 0:
        print(f"🎉 GIFs saved to: {OUTPUT_DIR}")
        print("\n💡 Next steps:")
        print("   1. Check the downloaded GIFs")
        print("   2. They should automatically work in your workouts!")
        print("   3. Update src/utils/exerciseMedia.js if needed\n")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n⚠️  Download cancelled by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Error: {e}")
        sys.exit(1)

File: scripts/downloadGifs.js
#!/usr/bin/env node

/**
 * Exercise GIF Downloader Script
 * 
 * Download exercise GIFs from URLs and save them with proper naming
 * 
 * Usage:
 *   node downloadGifs.js
 * 
 * Then follow the prompts or edit the exerciseUrls array below
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');

// ===== CONFIGURATION =====
// Add your GIF URLs here with exercise names
const exerciseUrls = [
  // Example format:
  // { name: 'Bench Press', url: 'https://example.com/bench-press.gif' },
  // { name: 'Squat', url: 'https://example.com/squat.gif' },
  // { name: 'Deadlift', url: 'https://example.com/deadlift.gif' },
  
  // Add your URLs below:
  
];

// Output directory (relative to project root)
const OUTPUT_DIR = path.join(__dirname, '..', '..', 'public', 'exercise-gifs');

// ===== HELPER FUNCTIONS =====

/**
 * Convert exercise name to kebab-case filename
 */
function toKebabCase(str) {
  return str
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

/**
 * Download a file from URL
 */
function downloadFile(url, filepath) {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith('https') ? https : http;
    
    const file = fs.createWriteStream(filepath);
    
    protocol.get(url, (response) => {
      // Handle redirects
      if (response.statusCode === 301 || response.statusCode === 302) {
        file.close();
        fs.unlinkSync(filepath);
        return downloadFile(response.headers.location, filepath)
          .then(resolve)
          .catch(reject);
      }
      
      if (response.statusCode !== 200) {
        file.close();
        fs.unlinkSync(filepath);
        return reject(new Error(`Failed to download: ${response.statusCode} ${response.statusMessage}`));
      }
      
      response.pipe(file);
      
      file.on('finish', () => {
        file.close();
        resolve();
      });
    }).on('error', (err) => {
      file.close();
      fs.unlinkSync(filepath);
      reject(err);
    });
  });
}

/**
 * Main download function
 */
async function downloadGifs() {
  console.log('🎬 Exercise GIF Downloader\n');
  
  // Ensure output directory exists
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    console.log(`✅ Created directory: ${OUTPUT_DIR}\n`);
  }
  
  if (exerciseUrls.length === 0) {
    console.log('⚠️  No URLs configured!');
    console.log('\nPlease edit this file and add URLs to the exerciseUrls array.\n');
    console.log('Example:');
    console.log('  {');
    console.log('    name: "Bench Press",');
    console.log('    url: "https://example.com/bench-press.gif"');
    console.log('  },\n');
    return;
  }
  
  console.log(`📥 Downloading ${exerciseUrls.length} exercise GIFs...\n`);
  
  let successCount = 0;
  let failCount = 0;
  
  for (const exercise of exerciseUrls) {
    const filename = `${toKebabCase(exercise.name)}.gif`;
    const filepath = path.join(OUTPUT_DIR, filename);
    
    try {
      process.stdout.write(`  Downloading "${exercise.name}"... `);
      await downloadFile(exercise.url, filepath);
      
      const stats = fs.statSync(filepath);
      const sizeMB = (stats.size / (1024 * 1024)).toFixed(2);
      
      console.log(`✅ (${sizeMB} MB)`);
      successCount++;
      
      // Small delay to be nice to servers
      await new Promise(resolve => setTimeout(resolve, 500));
      
    } catch (error) {
      console.log(`❌ Failed: ${error.message}`);
      failCount++;
    }
  }
  
  console.log('\n' + '='.repeat(50));
  console.log(`✅ Success: ${successCount}`);
  if (failCount > 0) {
    console.log(`❌ Failed: ${failCount}`);
  }
  console.log('='.repeat(50) + '\n');
  
  if (successCount > 0) {
    console.log('🎉 GIFs saved to:', OUTPUT_DIR);
    console.log('\n💡 Next steps:');
    console.log('   1. Check the downloaded GIFs');
    console.log('   2. They should automatically work in your workouts!');
    console.log('   3. Update src/utils/exerciseMedia.js if needed\n');
  }
}

// Run the script
if (require.main === module) {
  downloadGifs().catch(error => {
    console.error('❌ Error:', error.message);
    process.exit(1);
  });
}

module.exports = { downloadGifs, toKebabCase };
